azext_edge/edge/util/file_operations.py

```diff
@@ -0,0 +1,135 @@
+def dump_content_to_file(
+    content: List[dict],
+    file_name: str,
+    extension: str,
+    fieldnames: Optional[List[str]] = None,
+    output_dir: Optional[str] = None,
+    replace: bool = False
+) -> PurePath:
+    output_dir = normalize_dir(output_dir)
+    file_path = os.path.join(output_dir, f"{file_name}.{extension}")
+    if os.path.exists(file_path):
+        if not replace:
+            raise FileExistsError(f"File {file_path} already exists. Please choose another file name or add replace.")
+        logger.warning(f"The file {file_path} will be overwritten.")
+    if extension.endswith("csv"):
+        with open(file_path, "w", newline="", encoding="utf-8") as f:
+            if not fieldnames:
+                fieldnames = content[0].keys()
+            writer = csv.DictWriter(f, fieldnames=fieldnames)
+            writer.writeheader()
+            writer.writerows(content)
+        return file_path
+
+    # These let you dump to a string before writing to file
+    if extension == "json":
+        content = json.dumps(content, indent=2)
+    elif extension in ["yaml", "yml"]:
+        content = yaml.dump(content)
+    with open(file_path, "w", encoding="utf-8") as f:
+        f.write(content)
+
+    return file_path
+
+
+def normalize_dir(dir_path: Optional[str] = None) -> PurePath:
+    if not dir_path:
+        dir_path = "."
+    if "~" in dir_path:
+        dir_path = os.path.expanduser(dir_path)
+    dir_path = os.path.abspath(dir_path)
+    dir_pure_path = PurePath(dir_path)
+    if not os.path.exists(str(dir_pure_path)):
+        os.makedirs(dir_pure_path, exist_ok=True)
+
+    return dir_pure_path
+
+
+def read_file_content(file_path: str, read_as_binary: bool = False) -> Union[bytes, str]:
+    from pathlib import Path
+
+    logger.debug("Processing %s", file_path)
+    pure_path = Path(os.path.abspath(os.path.expanduser(file_path)))
+
+    if not pure_path.exists():
+        raise FileOperationError(f"{file_path} does not exist.")
+
+    if not pure_path.is_file():
+        raise FileOperationError(f"{file_path} is not a file.")
+
+    if read_as_binary:
+        logger.debug("Reading %s as binary", file_path)
+        return pure_path.read_bytes()
+
+    # Try with 'utf-8-sig' first, so that BOM in WinOS won't cause trouble.
+    for encoding in ["utf-8-sig", "utf-8"]:
+        try:
+            logger.debug("Reading %s as %s", file_path, encoding)
+            return pure_path.read_text(encoding=encoding)
+        except (UnicodeError, UnicodeDecodeError):
+            pass
+
+    raise FileOperationError(f"Failed to decode file {file_path}.")
+
+
+def read_file_content_as_dict(file_path: str) -> Union[Iterable, dict]:
+    extension = file_path.split(".")[-1]
+    invalid_extension = extension not in ["json", "yaml", "yml", "csv"]
+    content = read_file_content(file_path)
+    result: Optional[Union[Iterable, dict]] = None
+    if invalid_extension or extension == "json":
+        result = _try_loading_as(
```

Comment:
In this commit _try_loading_as still always casts to list, so anything that uses it returns list...when is dict returned?

Reply:
fixed